src/asm/parser.y

```diff
@@ -1194,6 +1196,40 @@ ds		: T_POP_DS uconst { sect_Skip($2, true); }
 			sect_RelBytes($2, $4.args, $4.nbArgs);
 			freeDsArgList(&$4);
 		}
+		| T_POP_DS ds_align trailing_comma {
+			uint32_t n = sect_GetAlignBytes($2.alignment, $2.alignOfs);
+
+			sect_Skip(n, true);
+			sect_AlignPC($2.alignment, $2.alignOfs);
+		}
+		| T_POP_DS ds_align T_COMMA ds_args trailing_comma {
+			uint32_t n = sect_GetAlignBytes($2.alignment, $2.alignOfs);
+
+			sect_RelBytes(n, $4.args, $4.nbArgs);
+			sect_AlignPC($2.alignment, $2.alignOfs);
+			freeDsArgList(&$4);
+		}
+;
+
+ds_align	: T_OP_ALIGN T_LBRACK uconst T_RBRACK {
+			if ($3 > 16) {
+				error("Alignment must be between 0 and 16, not %u\n", $3);
+			} else {
+				$$.alignment = $3;
+				$$.alignOfs = 0;
+			}
+		}
+		| T_OP_ALIGN T_LBRACK uconst T_COMMA uconst T_RBRACK {
+			if ($3 > 16) {
+				error("Alignment must be between 0 and 16, not %u\n", $3);
+			} else if ($5 >= 1 << $3) {
+				error("Offset must be between 0 and %u, not %u\n",
```

Comment:
Negative offsets should be allowed (just normalized), because people do use them occasionally. In fact, the constraints for the offset are very similar to the constraints for `-Wtruncation=1`, so why not use the same boundaries for a warning? (And this should be an enabled-by-default warning, not an error — an alignment of `4, 20` is equivalent to an alignment of `4, 4`, so this shouldn't be an error because it can be interpreted unambiguously.)

Reply:
Negative offsets have not yet been allowed. That can be a separate issue.

Reply:
Are they not allowed there either? If so, then yes, that could be.

Reply:
Filed as #1248.